src/Microsoft.ML.Transforms/CategoricalHashTransform.cs

```diff
@@ -120,6 +120,41 @@ public sealed class Arguments : TransformInputBase
 
         public const string UserName = "Categorical Hash Transform";
 
+        public static IDataTransform Create(IHostEnvironment env, IDataView input, int hashBits = 16, uint seed = 314489979,
+            bool ordered = true, int invertHash = 0, CategoricalTransform.OutputKind outputKind = CategoricalTransform.OutputKind.Bag, 
+            params string[] inputColumns)
+        {
+            var inputOutputColumns = new(string inputColumn, string outputColumn)[inputColumns.Length];
+            for (int i = 0; i < inputColumns.Length; i++)
+            {
+                inputOutputColumns[i].inputColumn = inputOutputColumns[i].outputColumn = inputColumns[i];
+            }
+            return Create(env, input, hashBits, seed, ordered, invertHash, outputKind, inputOutputColumns);
+        }
+
+        public static IDataTransform Create(IHostEnvironment env, IDataView input, int hashBits = 16, uint seed = 314489979, 
+            bool ordered = true, int invertHash = 0, CategoricalTransform.OutputKind outputKind = CategoricalTransform.OutputKind.Bag, 
+            params(string inputColumn, string outputColumn)[] inputOutputColumns)
```

Comment:
>params(string inputColumn, string outputColumn)[] inputOutputColumns [](start = 12, length = 68)

Let's try not to do this "multi-column" thing where reasonably we cannot. I'd like to have each such transform like this handle the *single* mapping case easily -- the multi-column thing, if necessary, I'm generally happy to tell people to revert to the `Arguments` scheme.

By insisting on handling that, and doing this `params` thing, you have introduced something incredibly unfortunate: by far the *most* important argument is the name of the output (and, if different, the source), but since it is a `params` it must go *last*. That's really too bad!

I propose that all constructor signatures for all one-to-one transforms follow this scheme:

```csharp
public FooTransform(IHostEnvironment env, IDataView input, string name, string source = null, ...)
```

Where of course, `...` refers to any arguments.

Reply:
It is fine and expected that the `Arguments` case will fulfill some cases that the convenience constructor cannot. The convenience constructor should handle the most *common* cases of the transform successfully, and with as much brevity as possible. (Indeed, in the case of the GUI, it *cannot* handle multiple columns in one go, yet most users seem perfectly happy with this. I expect the same will be true of this API.) The constructor certainly needn't cover the total, full breadth of everything in the `Arguments`. For example, I can definitely see exposing `hashBits` and `invertHash`, but I'm not sure I've seen too many people set `seed` or `ordered`.

---
In reply to: [197980521](https://github.com/dotnet/machinelearning/pull/405#discussion_r197980521) [](ancestors = 197980521)